File: src/lib/scala/Slate.Core/src/main/scala/slate/core/cloud/CloudFilesBase.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.core.cloud

import java.io.{InputStreamReader, BufferedReader, InputStream, ByteArrayInputStream}

import slate.common.{IO, Result, Files}

/**
  * Abstraction for cloud base file storage and retrieval.
  */
abstract class CloudFilesBase (val _defaultFolder:String, val _createDefaultFolder:Boolean) extends CloudActions {

  def connect(args:Any): Unit = { }


  def connectWith(key:String, pass:String, tag:String):Unit = { }


  def create(name:String, content:String): Unit =
  {
    create(_defaultFolder, name, content)
  }


  def createFromPath(name:String, filePath:String): Unit =
  {
    val content = loadFromFile(filePath)
    create(_defaultFolder, name, content)
  }


  def createFromPath(folder:String, name:String, filePath:String): Result[String] =
  {
    //val content = "simulating from file : " + filePath
    val content = loadFromFile(filePath)
    create(folder, name, content)
  }


  def delete(name:String): Unit =
  {
    delete(_defaultFolder, name)
  }


  def getAsText(name:String): Result[String] =
  {
    getAsText(_defaultFolder, name)
  }


  def download(name:String, localFolder:String): Unit =
  {
    download(_defaultFolder, name, localFolder)
  }


  def update(name:String, content:String): Unit =
  {
    update(_defaultFolder, name, content)
  }


  def updateFromPath(name:String, filePath:String): Result[String] =
  {
    val content = "simulating from file : " + filePath; //loadFromFile(filePath)
    update(_defaultFolder, name, content)
  }


  def updateFromPath(folder:String, name:String, filePath:String): Result[String] =
  {
    val content = loadFromFile(filePath)
    update(folder, name, content)
  }


  def createRootFolder(rootFolder:String):Unit


  def create(folder:String, name:String, content:String):Result[String]


  def update(folder:String, name:String, content:String): Result[String]


  def delete(folder:String, name:String):Result[String]


  def getAsText(folder:String, name:String):Result[String]


  def download(folder:String, name:String, localFolder:String): Result[String]


  def downloadToFile(folder:String, name:String, filePath:String):Result[String]


  protected def loadFromFile(filePath:String):String =
  {
    Files.readAllText(filePath)
  }


  protected def toInputStream(content:String):InputStream =
  {
    new ByteArrayInputStream(content.getBytes())
  }


  protected def toString(input: InputStream): String = {
    scala.io.Source.fromInputStream(input).getLines().mkString
  }
}

File: src/lib/kotlin/slatekit-core/src/main/kotlin/slatekit/core/cloud/CloudFilesBase.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.core.cloud

import slatekit.common.ResultEx
import java.io.ByteArrayInputStream
import java.io.File
import java.io.InputStream

/**
 * Abstraction for cloud base file storage and retrieval.
 */
abstract class CloudFilesBase(val _defaultFolder: String, val _createDefaultFolder: Boolean) : CloudActions {


    /**
     * hook for any initialization
     */
    open fun init(): Unit {

    }


    fun create(name: String, content: String) {
        create(_defaultFolder, name, content)
    }


    fun createFromPath(name: String, filePath: String) {

        val content = loadFromFile(filePath)
        create(_defaultFolder, name, content)
    }


    fun createFromPath(folder: String, name: String, filePath: String): ResultEx<String> {
        //val content = "simulating from file : " + filePath
        val content = loadFromFile(filePath)
        return create(folder, name, content)
    }


    fun delete(name: String): Unit {
        delete(_defaultFolder, name)
    }


    fun getAsText(name: String): ResultEx<String> = getAsText(_defaultFolder, name)


    fun download(name: String, localFolder: String): ResultEx<String> {
        return download(_defaultFolder, name, localFolder)
    }


    fun downloadToFile(name: String, localFilePath: String): ResultEx<String> {
        return downloadToFile(_defaultFolder, name, localFilePath)
    }


    fun update(name: String, content: String): Unit {
        update(_defaultFolder, name, content)
    }


    fun updateFromPath(name: String, filePath: String): ResultEx<String> {
        val content = "simulating from file : " + filePath; //loadFromFile(filePath)
        return update(_defaultFolder, name, content)
    }


    fun updateFromPath(folder: String, name: String, filePath: String): ResultEx<String> {
        val content = loadFromFile(filePath)
        return update(folder, name, content)
    }


    abstract fun createRootFolder(rootFolder: String): Unit


    abstract fun create(folder: String, name: String, content: String): ResultEx<String>


    abstract fun update(folder: String, name: String, content: String): ResultEx<String>


    abstract fun delete(folder: String, name: String): ResultEx<String>


    abstract fun getAsText(folder: String, name: String): ResultEx<String>


    abstract fun download(folder: String, name: String, localFolder: String): ResultEx<String>


    abstract fun downloadToFile(folder: String, name: String, filePath: String): ResultEx<String>


    protected fun loadFromFile(filePath: String): String {
        return File(filePath).readText()
    }


    protected fun toInputStream(content: String): InputStream {
        return ByteArrayInputStream(content.toByteArray())
    }


    protected fun toString(input: InputStream): String {
        return input.bufferedReader().use { it.readText() }
    }
}
